### PR TITLE
Feature: Add autoupdate feature and modify package.json to support autoupdate

### PR DESCRIPTION
**Pull request by jane-rose**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -116,6 +116,7 @@
       "target": [
         "nsis"
       ],
+      "verifyUpdateCodeSignature": false,
```
Comment (parachvte) [inline package.json:119]: revert to true?

File: package.json [hunk 2]
```diff
@@ -177,6 +178,7 @@
     "debug": "~3.1.0",
     "detect-browser": "~3.0.0",
     "electron-store": "~5.1.1",
+    "electron-updater": "^4.3.5",
```
Comment (parachvte) [inline package.json:181]: Use ~ tilde versioning

File: src/app/containers/App.jsx [hunk 3]
```diff
@@ -64,7 +64,12 @@ class App extends PureComponent {
         onRecovery: PropTypes.func.isRequired,
         quitRecovery: PropTypes.func.isRequired,
         saveAsFile: PropTypes.func.isRequired,
-        openProject: PropTypes.func.isRequired
+        openProject: PropTypes.func.isRequired,
+        updateStartDownload: PropTypes.func.isRequired,
+        updateCheckForUpdateImmediately: PropTypes.func.isRequired,
+        shouldCheckForUpdate: PropTypes.bool.isRequired,
+        isStartDownload: PropTypes.bool.isRequired,
```
Comment (parachvte) [inline src/app/containers/App.jsx:71]: Bad naming

File: src/app/containers/App.jsx [hunk 4]
```diff
@@ -64,7 +64,12 @@ class App extends PureComponent {
         onRecovery: PropTypes.func.isRequired,
         quitRecovery: PropTypes.func.isRequired,
         saveAsFile: PropTypes.func.isRequired,
-        openProject: PropTypes.func.isRequired
+        openProject: PropTypes.func.isRequired,
+        updateStartDownload: PropTypes.func.isRequired,
+        updateCheckForUpdateImmediately: PropTypes.func.isRequired,
+        shouldCheckForUpdate: PropTypes.bool.isRequired,
```
Comment (parachvte) [inline src/app/containers/App.jsx:70]: ?

File: src/app/containers/App.jsx [hunk 5]
```diff
@@ -241,6 +267,12 @@ class App extends PureComponent {
             UniApi.Menu.setItemEnabled('save-as', !!headType);
             UniApi.Menu.setItemEnabled('save', !!headType);
         }
+        if (nextProps.checkForUpdateImmediately && nextProps.isStartDownload) {
```
Comment (parachvte) [inline src/app/containers/App.jsx:270]: ???

File: src/app/containers/Settings/General/index.styl [hunk 6]
```diff
@@ -30,3 +30,19 @@
 
     .about-product-description
         margin-bottom: 10px
+.autoupdate-wrapper{
```
Comment (parachvte) [inline src/app/containers/Settings/General/index.styl:33]: spaces?

File: src/app/flux/machine/index.js [hunk 7]
```diff
@@ -115,7 +115,15 @@ const INITIAL_STATE = {
...
     },
     shouldShowCncWarning: true,
-    connectionTimeout: 3000
+    connectionTimeout: 3000,
+
+    // autoUpdate
+    // Whether to check for available updates when the software is opened
+    shouldCheckForUpdate: true,
+    // Whether an update is being downloaded
+    isStartDownload: false,
```
Comment (parachvte) [inline src/app/flux/machine/index.js:124]: bad naming

File: src/app/lib/uni-api.js [hunk 8]
```diff
@@ -15,6 +15,78 @@ const Event = {
...
 
+/**
+ *  Update control in electron
+ */
+const Update = {
+    checkForUpdate(shouldCheckForUpdate) {
+        if (isElectron()) {
+            const { ipcRenderer } = window.require('electron');
+            if (shouldCheckForUpdate) {
+                ipcRenderer.send('checkForUpdate');
```
Comment (parachvte) [inline src/app/lib/uni-api.js:26]: 1. Use common prefix for the event keys, like "auto-update:check", "auto-update:download" to group events together
2. Use constants for event keys

File: src/app/lib/uni-api.js [hunk 9]
```diff
@@ -15,6 +15,78 @@ const Event = {
...
+                defaultId: 1,
+                title: i18n._('Update Snapmaker Luban'),
+                message: i18n._(`Snapmaker Luban ${releaseName} Update`),
+                detail: i18n._(`Current version : ${oldVersionn}`)
+                // detail: 'A new version has been detected. Should i download it now?'
+            };
+
+            dialog.showMessageBox(dialogOpts).then((returnValue) => {
+                if (returnValue.response === 1) {
+                    ipcRenderer.send('isDownloadNow');
```
Comment (parachvte) [inline src/app/lib/uni-api.js:48]: bad event naming

File: src/app/lib/uni-api.js [hunk 10]
```diff
@@ -15,6 +15,78 @@ const Event = {
...
+                }
+            });
+        }
+    },
+    haveStartedDownload() {
```
Comment (parachvte) [inline src/app/lib/uni-api.js:53]: 1. has
2. `downloadHasStarted`

File: src/main.js [hunk 11]
```diff
@@ -76,6 +77,61 @@ function getBrowserWindowOptions() {
...
+    // Emitted when there is no available update.
+    autoUpdater.on('update-not-available', () => {
+        sendUpdateMessage(message.updateNotAva);
+    });
+    autoUpdater.on('download-progress', (progressObj) => {
+        mainWindow.setProgressBar(progressObj.percent / 100);
+    });
+    // downloadInfo — for generic and github providers
+    autoUpdater.on('update-downloaded', (downloadInfo) => {
+        ipcMain.on('isUpdateNow', () => {
```
Comment (parachvte) [inline src/main.js:118]: ???

File: src/app/containers/App.jsx [hunk 12]
```diff
@@ -256,7 +246,9 @@ class App extends PureComponent {
         this.actions.initFileOpen();
         // auto update
         setTimeout(() => {
-            this.actions.checkForUpdate(this.props.shouldCheckForUpdate);
+            if (this.props.shouldCheckForUpdate) {
```
Comment (parachvte) [inline src/app/containers/App.jsx:249]: put condition outside of `setTimeout`?

File: src/app/containers/App.jsx [hunk 13]
```diff
@@ -139,26 +136,19 @@ class App extends PureComponent {
...
-                this.props.updateStartDownload(true);
+            UniApi.Event.on('download-has-started', () => {
+                this.props.updateIsDownloading(true);
             });
-            UniApi.Event.on('updateAvailable', (event, downloadInfo, oldVersionn) => {
+            UniApi.Event.on('update-available', (event, downloadInfo, oldVersionn) => {
                 UniApi.Update.downloadUpdate(downloadInfo, oldVersionn);
             });
-            UniApi.Event.on('isUpdateNow', (event, downloadInfo) => {
+            UniApi.Event.on('is-update-now', (event, downloadInfo) => {
```
Comment (parachvte) [inline src/app/containers/App.jsx:149]: 1. What does `is-update-now` mean, updating app / downloading app / replacing app?
2. `is-updating-now`?

File: src/app/containers/App.jsx (hunk 12 above)
Comment (jane-rose) [reply src/app/containers/App.jsx:249]: no, it does not work

File: src/app/containers/App.jsx (hunk 13 above)
Comment (jane-rose) [reply src/app/containers/App.jsx:149]: replacing app, change it to is-replacing-app-now

**Pull request merged**